src/arch/wasm/CodeGen.zig

```diff
@@ -6552,19 +6578,68 @@ fn divSigned(func: *CodeGen, lhs: WValue, rhs: WValue, ty: Type) InnerError!WVal
     return result;
 }
 
-/// Retrieves the absolute value of a signed integer
-/// NOTE: Leaves the result value on the stack.
-fn signAbsValue(func: *CodeGen, operand: WValue, ty: Type) InnerError!WValue {
+/// Remainder after floor division, defined by:
+/// @divFloor(a, b) * b + @mod(a, b) = a
+fn airMod(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
+    const bin_op = func.air.instructions.items(.data)[inst].bin_op;
+
+    const mod = func.bin_file.base.options.module.?;
+    const ty = func.typeOfIndex(inst);
+    const lhs = try func.resolveInst(bin_op.lhs);
+    const rhs = try func.resolveInst(bin_op.rhs);
+
+    if (ty.isUnsignedInt(mod)) {
+        _ = try func.binOp(lhs, rhs, ty, .rem);
+    } else if (ty.isSignedInt(mod)) {
+        // The wasm rem instruction gives the remainder after truncating division (rounding towards
+        // 0), equivalent to @rem.
+        // We make use of the fact that:
+        // @mod(a, b) = @rem(@rem(a, b) + b, b)
+        const int_bits = ty.intInfo(mod).bits;
+        const wasm_bits = toWasmBits(int_bits) orelse {
+            return func.fail("TODO: `@mod` for signed integers larger than 64 bits ({d} bits requested)", .{int_bits});
+        };
+
+        if (wasm_bits > 64) {
+            return func.fail("TODO: `@mod` for signed integers larger than 64 bits ({d} bits requested)", .{int_bits});
+        }
+
+        const lhs_wasm = if (wasm_bits != int_bits)
+            try (try func.signExtendInt(lhs, ty)).toLocal(func, ty)
+        else
+            lhs;
+
+        const rhs_wasm = if (wasm_bits != int_bits)
+            try (try func.signExtendInt(rhs, ty)).toLocal(func, ty)
+        else
+            rhs;
+
+        _ = try func.binOp(lhs_wasm, rhs_wasm, ty, .rem);
+        _ = try func.binOp(.stack, rhs_wasm, ty, .add);
+        _ = try func.binOp(.stack, rhs_wasm, ty, .rem);
+    } else {
+        return func.fail("TODO: implement `@mod` on floating point types for {}", .{func.target.cpu.arch});
+    }
+
+    const result = try func.allocLocal(ty);
+    try func.addLabel(.local_set, result.local.value);
+    func.finishAir(inst, result, &.{ bin_op.lhs, bin_op.rhs });
+}
+
+/// Sign extends an N bit signed integer and pushes the result to the stack.
+/// The result will be sign extended to 32 bits if N <= 32 or 64 bits if N <= 64.
+/// Support for integers wider than 64 bits has not yet been implemented.
+fn signExtendInt(func: *CodeGen, operand: WValue, ty: Type) InnerError!WValue {
```

Comment:
Would you mind updating the doc comment to tell the developer it may leave the value on the stack, just like the old comment did?

Reply:
The new comment states that the result is pushed to the stack - perhaps you missed this?

Reply:
You're absolutely right - My bad!